### PR TITLE
Add CRC based podified EDPM zuul job

### PR DESCRIPTION
**Pull request by raukadah**
The job runs on centos-9 24 GB ram crc node provided by the RDO infra.

When the job starts the crc will be started and `make openstack` command will be executed in the pre-step of base-crc job due to pull secret dependency.

pre_pull_images depends on pull_secrets, also get pulled in the pre-step.

This jobs does the following things:
- Start crc node "--memory 16000 --disk-size 120 --cpus 6"
- Run Make openstack and pull pre_pulled_images needed for deployment
- Create local storage
- Deploy podified control plane
- Deletes nova scheduler and list compute services
- Creates the EDPM node
- Add repos on the edpm node
- Perform edpm-deploy
- Runs validation

Depends-On: https://github.com/openstack-k8s-operators/install_yamls/pull/194
Depends-On: https://github.com/openstack-k8s-operators/install_yamls/pull/193

### CODE REVIEW

File: zuul.d/jobs.yaml [hunk 1]
```diff
@@ -0,0 +1,45 @@
+---
+- job:
+    name: base-crc-openstack
+    parent: base-crc
+    timeout: 10800
+    nodeset: centos-8-crc-single-node-centos-9-external-compute
```
Comment (cjeanner) [inline zuul.d/jobs.yaml:6]: note, we now have a CRC node running on cs9 - it's working just fine, used in ci-framework job. Maybe something to consider?

**Comment by raukadah** [comment]
recheck

**Comment by fao89** [comment]
FYI: https://github.com/rdo-infra/rdo-jobs/pull/2

**Comment by raukadah** [comment]
recheck

**Review by fao89 — COMMENTED** [review]
it is looking good!

**Comment by raukadah** [comment]
recheck

File: ci/scripts/validate_edpm.sh [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+openstack router add subnet priv_router priv_sub
+openstack router set priv_router --external-gateway public
+
+openstack server create --flavor m1.small --image cirros --nic net-id=private test
+openstack floating ip create public --floating-ip-address 192.168.122.20
+openstack server add floating ip test 192.168.122.20
+openstack server list
+openstack security group rule create --protocol icmp --ingress --icmp-type -1 $(openstack security group list --project admin -f value -c ID)
+openstack security group rule create --protocol tcp --ingress --dst-port 22 $(openstack security group list --project admin -f value -c ID)
+ping -c4 192.168.122.20
```
Comment (fao89) [inline ci/scripts/validate_edpm.sh:28]: sweet!

**Review by fao89 — APPROVED** [review]

**Comment by raukadah** [comment]
recheck

**Comment by raukadah** [comment]
https://logserver.rdoproject.org/62/162/54b8561ff13288cf810d4d13f6e059928396c193/github-check/centos-9-crc-singlenode-podified-edpm-deployment/c020f02/job-output.txt
```
openstack flavor create --ram 512 --vcpus 1 --disk 1 --ephemeral 1 m1.small
2023-03-31 10:00:07.841529 | controller | Unable to establish connection to http://keystone-public-openstack.apps-crc.testing/v3/auth/tokens: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))
```

**Comment by raukadah** [comment]
recheck

File: ci/playbooks/deploy_podified_openstack.yaml [hunk 3]
```diff
@@ -55,3 +59,10 @@
         oc get secrets;
         oc rsh openstackclient openstack compute service list;
         oc rsh openstackclient openstack network agent list;
+<<<<<<< HEAD
```
Comment (SeanMooney) [inline ci/playbooks/deploy_podified_openstack.yaml:62]: you have an unresolved merge conflict here

Comment (raukadah) [reply ci/playbooks/deploy_podified_openstack.yaml]: Done

File: ci/scripts/validate_edpm.sh [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+openstack subnet create priv_sub --subnet-range 192.168.0.0/24 --network private
+openstack network create public --external  --provider-network-type flat --provider-physical-network datacentre
+openstack subnet create pub_sub --subnet-range 192.168.122.0/24 --allocation-pool start=192.168.122.200,end=192.168.122.210 --gateway 192.168.122.1 --no-dhcp --network public
+openstack router create priv_router
+openstack router add subnet priv_router priv_sub
+openstack router set priv_router --external-gateway public
+
+openstack server create --flavor m1.small --image cirros --nic net-id=private test
+openstack floating ip create public --floating-ip-address 192.168.122.20
+openstack server add floating ip test 192.168.122.20
```
Comment (slagle) [inline ci/scripts/validate_edpm.sh:24]: why isn't this IP in the pub_sub subnet range?

File: ci/scripts/validate_edpm.sh
Comment (slagle) [inline ci/scripts/validate_edpm.sh:1]: maybe this should be under devsetup/scripts? I think we should add a Makefile target to run it as well: `make edpm_deploy_instance`
that could be in a separate PR though

Comment (raukadah) [reply ci/scripts/validate_edpm.sh:1]: Make sense! Added it here https://github.com/openstack-k8s-operators/install_yamls/pull/194

**Comment by raukadah** [comment]
recheck

File: ci/playbooks/deploy_edpm.yaml [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        oc get openstackdataplane
+        oc get openstackdataplanerole
+        oc get openstackdataplanenode
+        oc get pods | grep edpm
+      ignore_errors: true
+
+    - name: Get dataplane-deployment-run pod name
+      register: edpm_pod
+      ansible.builtin.shell:
+        cmd: oc get pods -o name | grep -m1  dataplane-deployment-run-openstack
```
Comment (fultonj) [inline ci/playbooks/deploy_edpm.yaml:51]: So other `dataplane-deployment` AEE pods will run but you're waiting for `dataplane-deployment-run-openstack`, which is the last one as defined in deployment.go [1]. That is not the last AEE to run however, as when that finishes two nova playbooks run in parallel [2]. When that ansible is finished running (with two edpm compute nodes) it looks like this [3].

Maybe it's more correct, if the intent is to be sure all ansible jobs have run, to wait for the job matching `nova-edpm-compute-0-deploy-nova` to be finished instead of the one matching `dataplane-deployment-run-openstack`.

I see however, that we have green jobs so we don't have to block on this. It could lead to more stable CI however to be sure all relevant AEE jobs have run before running the validation. I suspect the green jobs worked but it's probably because they won the race.

[1] https://github.com/openstack-k8s-operators/dataplane-operator/blob/main/pkg/deployment/deployment.go#L329

[2] https://github.com/openstack-k8s-operators/nova-operator/tree/master/playbooks

[3]
```
[fultonj@osp-storage-01 devsetup]$ oc get pods | egrep "dataplane|edpm"
dataplane-deployment-configure-network-f7qpm-mj9dp                0/1     Completed                0                 9m24s
dataplane-deployment-configure-openstack-hnzwf-9prkt              0/1     Completed                0                 3m23s
dataplane-deployment-configure-os-x7g52-vw79w                     0/1     Completed                0                 5m44s
dataplane-deployment-install-openstack-pdr9v-d9bcv                0/1     Completed                0                 3m43s
dataplane-deployment-install-os-k5v9r-d94rb                       0/1     Completed                0                 6m35s
dataplane-deployment-run-openstack-kh5d8-hbxbr                    0/1     Completed                0                 2m57s
dataplane-deployment-run-os-jndpm-r5fxc                           0/1     Completed                0                 4m25s
dataplane-deployment-validate-network-kns2z-wdl2z                 0/1     Completed                0                 6m45s
nova-edpm-compute-0-deploy-libvirt-gh4hc                          0/1     Completed                0                 2m24s
nova-edpm-compute-0-deploy-nova-gxh7w                             0/1     Completed                0                 2m24s
nova-edpm-compute-1-deploy-libvirt-kxsf6                          0/1     Completed                0                 2m24s
nova-edpm-compute-1-deploy-nova-h5glm                             0/1     Completed                0                 2m24s
[fultonj@osp-storage-01 devsetup]$ 
```

File: ci/playbooks/deploy_edpm.yaml [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+    - name: Wait for dataplane-deployment-run pod name
+      register: deploy_status
+      ansible.builtin.shell:
+        cmd: "oc wait --timeout=0 --for=jsonpath='{.status.phase}'=Succeeded {{ edpm_pod.stdout |trim }}"
+      until: deploy_status.rc == 0
+      retries: 100
+      delay: 20
+
+    - name: Get the logs of EDPM Play
```
Comment (fultonj) [inline ci/playbooks/deploy_edpm.yaml:65]: You could get the logs of all EDPM plays [1] not just the `dataplane-deployment-run-openstack-*` one.

[1]  https://paste.opendev.org/show/bM8DkvYsfMaEX0vFhqMI/

**Review by fultonj — COMMENTED** [review]
I think this a big improvement from what we had. There's still some areas for improvement however which I have noted below. I'm OK with merging if there can be a follow up to address them.

File: ci/playbooks/deploy_edpm.yaml (hunk 5 above)
Comment (fultonj) [reply ci/playbooks/deploy_edpm.yaml:51]: For example...

`for POD in $(oc get pods -o name | egrep "dataplane|edpm"); do echo $POD; oc logs $POD; done`

If one of the many playbooks fail, we'll want their logs.

File: ci/playbooks/deploy_edpm.yaml (hunk 6 above)
Comment (fultonj) [reply ci/playbooks/deploy_edpm.yaml:65]: For example...

`for POD in $(oc get pods -o name | egrep "dataplane|edpm"); do echo $POD; oc logs $POD; done`

If one of the many playbooks fail, we'll want their logs.

File: ci/playbooks/deploy_edpm.yaml (hunk 5 above)
Comment (raukadah) [reply ci/playbooks/deploy_edpm.yaml]: Thank you @fultonj for the nice suggestion, I have made the changes.

File: ci/playbooks/deploy_edpm.yaml (hunk 6 above)
Comment (raukadah) [reply ci/playbooks/deploy_edpm.yaml:65]: Done

**Review by slagle — APPROVED** [review]
